devito/arguments.py

```diff
@@ -158,198 +139,381 @@ class PtrArgument(Argument):
+
+        kwargs = self._offset_adjust(kwargs)
+
+        kwargs = self._extract_children_of_composites(kwargs)
+
+        values = self._derive_values(kwargs)
+
+        # The following is only being done to update the autotune flag. The actual value
+        # derivation for the dle arguments has moved inside the above _derive_values
+        # method.
+        # TODO: Refactor so this is not required
+        dim_sizes = dict([(d.name, runtime_dim_extent(d, values))
+                          for d in self.dimensions])
+        dle_arguments, dle_autotune = self._dle_arguments(dim_sizes)
+
+        assert(self._verify(values))
+
+        arguments = OrderedDict([(k.name, v) for k, v in values.items()])
+
+        return arguments, user_autotune and dle_autotune
+
+    def _offset_adjust(self, kwargs):
+        for k, v in kwargs.items():
+            if k in self.offsets:
+                kwargs[k] = v + self.offsets[k]
+        return kwargs
+
+    def _build_argument_mapper(self, parameters):
+        # Pass through SymbolicFunction
+        symbolic_functions = [x for x in parameters if isinstance(x, SymbolicFunction)]
+        dim_dep_mapper = OrderedDict()  # Mapper for dependencies between dimensions
+        tensor_arguments = []
+        for f in symbolic_functions:
+            argument = ArgumentVisitor().visit(f)
+            tensor_arguments.append(argument)
+            for i, d in enumerate(f.indices):
+                dim_dep_mapper.setdefault(d, []).append(Dependency("gets_value_from",
+                                                                   argument, param=i))
+
+        for arg in self.dle_arguments:
+            d = arg.argument
+            dim_dep_mapper.setdefault(d, []).append(Dependency("gets_value_from",
+                                                               derive_dle_argument_value,
+                                                               param=arg))
+
+        # Record dependencies in Dimensions
+        dim_param_mapper = OrderedDict([(k, DimensionParameter(k, v)) for k, v in
+                                        dim_dep_mapper.items()])
+
+        # Dimensions that are in parameters but not directly referenced in the expressions
+        more = dict([(x, DimensionParameter(x, [])) for x in self.dimensions if x not in
+                     dim_dep_mapper])
+        dim_param_mapper.update(more)
+
+        for dim in [x for x in self.dimensions if x.is_Stepping]:
+            dim_param_mapper[dim.parent].dependencies.append(
+                Dependency(Dependency.GETS_VALUE_FROM, dim_param_mapper[dim]))
+
+        dimension_parameters = list(dim_param_mapper.values())
+
+        scalar_arguments = flatten([ArgumentVisitor().visit(x)
+                                    for x in dimension_parameters])
+
+        other_arguments = [ArgumentVisitor().visit(x) for x in parameters
+                           if x not in tensor_arguments + scalar_arguments]
```

Comment:
With an ``OrderedDict`` (an ``OrderedSet``, TBH) you wouldn't need these extra checks, and you wouldn't need four different variables (``tensor_arguments``, ``dimension_parameters``, ...)

Reply:
I prefer to treat `tensor_arguments`, `dimension_parameters` etc separately. I don't see why that is wrong.

Reply:
sure, that was nitpicking on my side. However, let's remember this possibility *if* the performance of the arguments engine turns out to be poor, as I suspect an OrderedDict might make things much faster (as opposed to creating and concatenating lists, and using if-then-else).

Reply:
Premature optimisation?